### PR TITLE
Add overloads to CookieAuthentication to config options with services

### LINKED ISSUE #18772: DI support for auth handler options
https://github.com/dotnet/aspnetcore/issues/9337#issuecomment-581489114
https://github.com/dotnet/aspnetcore/issues/9337#issuecomment-581516787

### Describe the bug
The authentication handlers are configured in Startup.ConfigureServices and have trouble relying on services for any customization. E.g.
```
	.AddCookie(o =>
	{
		o.TicketDataFormat = new CustomDataFormat(settings.Secret, /* Add logging? */);
	});
```

IOptions supports DI for configure callbacks, but the existing overloads aren't compatible with the auth setup pattern.

### Workaround:
```
            services.AddAuthentication(options =>
            {
                options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
            }).AddCookie(CookieAuthenticationDefaults.AuthenticationScheme);

            services.AddTransient<IConfigureOptions<CookieAuthenticationOptions>>(sp =>
                new ConfigureNamedOptions<CookieAuthenticationOptions>(CookieAuthenticationDefaults.AuthenticationScheme, o =>
                {
                    o.TicketDataFormat = new CustomDataFormat("secret", sp.GetRequiredService<ILogger<CustomDataFormat>>());
                }));
```

### Proposals:
Some new overloads could make this easier. We should consider if these apply to all of the auth handlers, not just the Cookie samples given here.
```
.AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, (o, serviceProvider) =>
{
        o.TicketDataFormat = new CustomDataFormat("secret", sp.GetRequiredService<ILogger<CustomDataFormat>>());
});

.AddCookie<ILogger<CustomDataFormat>>(CookieAuthenticationDefaults.AuthenticationScheme, (o, logger) =>
{
        o.TicketDataFormat = new CustomDataFormat("secret", logger);
});

services.Configure<CookieAuthenticationOptions, ILogger<CustomDataFormat>>(CookieAuthenticationDefaults.AuthenticationScheme, o =>
{
       o.TicketDataFormat = new CustomDataFormat("secret", sp.GetRequiredService<ILogger<CustomDataFormat>>());
});
```

### PR DESCRIPTION
**Pull request by Kahbazi**
Contributes to #18772

### CODE REVIEW

**Comment by Tratcher** [comment]
Looks good so far. Can you update the cookie sample to DI inject the SessionStore?
https://github.com/dotnet/aspnetcore/blob/169f07c6675712aba5cf84c41e61838cebde2906/src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs#L22

**Comment by Tratcher** [comment]
@Kahbazi are you able to continue with this? or should we close it for now?

**Comment by anurse** [comment]
Triage: We're good with the API shape after this trial run. @Kahbazi can you move through the other Auth providers and update them the same way? We'd like to get them all in one PR.

**Comment by Tratcher** [comment]
The ref assemblies need to be updated. https://github.com/dotnet/aspnetcore/blob/master/docs/ReferenceAssemblies.md

**Comment by Kahbazi** [comment]
@Tratcher Should I also add Identity overloads in this PR?

> ConfigureApplicationCookie and related identity extensions are going to need something similar.

File: src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs [hunk 1]
```diff
@@ -14,12 +14,14 @@ public class Startup
...
         {
+            services.AddSingleton<MemoryCacheTicketStore>();
+
             // This can be removed after https://github.com/aspnet/IISIntegration/issues/371
             services.AddAuthentication(options =>
             {
                 options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                 options.DefaultChallengeScheme = CookieAuthenticationDefaults.AuthenticationScheme;
-            }).AddCookie(o => o.SessionStore = new MemoryCacheTicketStore());
+            }).AddCookie((o, services) => o.SessionStore = services.GetService<MemoryCacheTicketStore>());
```
Comment (Tratcher) [inline src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs]: I hate to second guess ourselves at this point, but what if this was `AddCookie<TService>(Action<CookieAuthOptions, TService))` so the above code could read like `AddCookie<MemoryCacheTicketStore>((o, store) => o.SessionStore = store))`. It's more elegant for the single service case (most common?), and then for multi-service you can inject IServiceProvider. Options already supports this, we'd only be changing our new overloads here.
@Kahbazi @anurse @HaoK thoughts?

**Comment by Tratcher** [comment]
> @Tratcher Should I also add Identity overloads in this PR?
> 
> > ConfigureApplicationCookie and related identity extensions are going to need something similar.

Yes, go ahead once we answer my other question about TService.

File: src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs (hunk 1 above)
Comment (HaoK) [reply src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs:24]: Yeah that has the benefit of looking just like the options builder apis, so its more consistent, I like that much better too

Comment (Kahbazi) [reply src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs:24]: Ok I will change the api. My only concern is users may not know about `IServiceProvider` and expect to pass two or more services as generic like Options. There should be a good comment for this api. So what comment should I add for `Action<CookieAuthOptions, TService)` parameter?

Comment (Tratcher) [reply src/Security/Authentication/Cookies/samples/CookieSessionSample/Startup.cs:24]: TService: A service resolved from the IServiceProvider for use when configuring this authentication provider. If you need multiple services then specify IServiceProvider and resolve them directly.

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks @Kahbazi

**Pull request merged**